iec/src/lowering/st.rs

```diff
@@ -0,0 +1,400 @@
+
+        let Variables {
+            local_variables,
+            parameters,
+        } = self.resolve_var_blocks(id, &function.var_blocks);
+        let return_ident = &function.return_type;
+        let return_type = match self.get_type_by_name(
+            &return_ident.value,
+            current_scope,
+            return_ident.loc(id),
+        ) {
+            Ok(t) => t,
+            Err(diag) => {
+                self.diags.push(diag);
+                return;
+            },
+        };
+        let function = Function {
+            local_variables,
+            parameters,
+            return_type,
+        };
+
+        let ent = self
+            .state
+            .entities
+            .build_entity()
+            .with(Name(name.to_string()), &mut self.state.names)
+            .with(location.clone(), &mut self.state.locations)
+            .with(ScopeRef(current_scope), &mut self.state.scope_refs)
+            .with(function, &mut self.state.functions)
+            .build();
+
+        self.add_global(name, Symbol::Function(ent), location);
+    }
+
+    fn resolve_var_blocks(
+        &mut self,
+        file_id: FileId,
+        blocks: &[syntax::VarBlock],
+    ) -> Variables {
+        let mut local_variables = Vec::new();
+        let mut return_types = Vec::new();
+        let mut parameters = Vec::new();
+
+        for block in blocks {
+            let dest = match block.kind {
+                syntax::VarBlockKind::Input => &mut parameters,
+                syntax::VarBlockKind::Output => &mut return_types,
+                syntax::VarBlockKind::Normal => &mut local_variables,
+                _ => unimplemented!(),
+            };
+
+            for decl in &block.declarations {
+                if let Some(ent) = self.resolve_var(file_id, decl) {
+                    dest.push(ent);
+                }
+            }
+        }
+
+        Variables {
+            local_variables,
+            parameters,
+        }
+    }
+
+    fn resolve_var(
+        &mut self,
+        file_id: FileId,
+        decl: &syntax::VariableDeclaration,
+    ) -> Option<Entity> {
+        let name = Name(decl.name.value.to_string());
+        let location = Location {
+            file: file_id,
+            span: decl.name.span,
+        };
+
+        let ty = match self.get_type_by_name(
+            &decl.declared_type.value,
+            self.global_scope,
+            location,
+        ) {
+            Ok(ty) => HasType { ty },
+            Err(diag) => {
+                self.diags.push(diag);
+                return None;
+            },
+        };
+
+        let ent = self
+            .state
+            .entities
+            .build_entity()
+            .with(name, &mut self.state.names)
+            .with(location, &mut self.state.locations)
+            .with(ty, &mut self.state.has_type)
+            .build();
+
+        Some(ent)
+    }
+}
+
+struct Variables {
```

Comment:
Where does globals go?

Reply:
That's a very good question... And I wish I had a good answer.

I'm not really sure what comprises a function signature in IEC languages. Is accessing a global variable just a function call at runtime, or should we make it part of the function signature itself? I mean "signature" in both senses, so the abstract data that is used by the compiler to reason about a function, as well as the concrete code generated (because WASM functions can only work with integers/floats, and I'm not sure if multiple returns is a thing).